package/libpiksi/libpiksi/src/loop.c

```diff
@@ -0,0 +1,467 @@
+ */
+static void pk_loop_destroy_uv_loop(uv_loop_t *uv_loop)
+{
+  if (uv_loop == NULL) {
+    return;
+  }
+  uv_walk(uv_loop, loop_destroy_callback, NULL);
+  while(uv_loop_alive(uv_loop)) {
+    if (uv_run(uv_loop, UV_RUN_NOWAIT) == 0) {
+      break;
+    }
+    piksi_log(LOG_DEBUG, "Re-running loop to close pending handles");
+  }
+  uv_loop_close(uv_loop);
+  free(uv_loop);
+}
+
+void pk_loop_destroy(pk_loop_t **pk_loop_loc)
+{
+  if (pk_loop_loc == NULL || *pk_loop_loc == NULL) {
+    return;
+  }
+
+  pk_loop_t *pk_loop = (pk_loop_t *)(*pk_loop_loc);
+  pk_loop_destroy_uv_handle((uv_handle_t *)pk_loop->timeout_timer);
+  pk_loop_destroy_uv_loop(pk_loop->uv_loop);
+  free(pk_loop);
+
+  *pk_loop_loc = NULL;
+}
+
+int pk_loop_get_signal_from_handle(void *handle)
+{
+  assert(handle != NULL);
+  assert(uv_handle_get_type((uv_handle_t *)handle) == UV_SIGNAL);
+  return ((uv_signal_t *)handle)->signum;
+}
+
+/**
+ * @brief signal_handler - wrapping callback for uv_signal_t
+ * @param signal: signal handle
+ * @param signum: signal that triggered this callback
+ */
+static void signal_handler(uv_signal_t *signal, int signum) {
+  uv_handle_t *handle = (uv_handle_t *)signal;
+  pk_loop_t *loop = pk_loop_from_uv_handle(handle);
+  pk_callback_ctx_t *cb_ctx = pk_callback_context_from_uv_handle(handle);
+  assert(signum == pk_loop_get_signal_from_handle(signal));
+
+  if (cb_ctx->callback != NULL) {
+    cb_ctx->callback(loop, handle, cb_ctx->data);
+  }
+}
+
+void * pk_loop_signal_handler_add(pk_loop_t *pk_loop,
+                                  int signal,
+                                  pk_loop_cb callback,
+                                  void *context)
+{
+  assert(pk_loop != NULL);
+
+  uv_signal_t *uv_signal = (uv_signal_t *)malloc(sizeof(uv_signal_t));
+  if (uv_signal == NULL) {
+    piksi_log(LOG_ERR, "Failed to allocate signal context");
+    goto failure;
+  }
+
+  if (uv_signal_init(pk_loop->uv_loop, uv_signal) != 0) {
+    piksi_log(LOG_ERR, "Failed to init signal context");
+    goto failure;
+  }
+
+  if (uv_signal_start(uv_signal, signal_handler, signal) != 0) {
+    piksi_log(LOG_ERR, "Failed to start signal context");
+    goto failure;
+  }
+
+  if (pk_loop_add_handle_context((uv_handle_t *)uv_signal, callback, context) != 0) {
+    piksi_log(LOG_ERR, "Failed to allocate callback context for signal handler add");
+    goto failure;
+  }
+
+  return uv_signal;
+
+failure:
+  pk_loop_destroy_uv_handle((uv_handle_t *)uv_signal);
+  return NULL;
+}
+
+/**
+ * @brief uv_loop_poll_handler - wrapping callback for uv_poll_t
+ * @param poller: poll handle
+ * @param status: poll operation status
+ * @param events: event that triggered this callback
+ */
+static void uv_loop_poll_handler(uv_poll_t *poller, int status, int events)
+{
+  uv_handle_t *handle = (uv_handle_t *)poller;
+  pk_loop_t *loop = pk_loop_from_uv_handle(handle);
+  pk_callback_ctx_t *cb_ctx = pk_callback_context_from_uv_handle(handle);
+
+  if (status < 0) {
```

Comment:
I haven't seen any examples of this error so far so it's hard to say. Looking through libuv's [error handling](http://docs.libuv.org/en/v1.x/errors.html) reference - it could be quite a few different things. At this point I thought a safe approach might be to simply log the error and skip that particular callback.